client/executor/wasmtime/src/instance_wrapper/globals_snapshot.rs

```diff
@@ -17,115 +17,62 @@
 /// immutable global variables always have the same values.
-pub struct GlobalsSnapshot {
-	handle: wasmtime_runtime::InstanceHandle,
-	preserved_mut_globals: Vec<(*mut wasmtime_runtime::VMGlobalDefinition, Value)>,
-}
+pub struct GlobalsSnapshot(Vec<SavedValue>);
 
 impl GlobalsSnapshot {
 	/// Take a snapshot of global variables for a given instance.
 	pub fn take(instance_wrapper: &InstanceWrapper) -> Result<Self> {
-		// EVIL:
-		// Usage of an undocumented function.
-		let handle = instance_wrapper.instance.handle().clone().handle;
-
-		let mut preserved_mut_globals = vec![];
-
-		for global_idx in instance_wrapper.imported_globals_count..instance_wrapper.globals_count {
-			let (def, global) = match handle.lookup_by_declaration(
-				&wasmtime_environ::EntityIndex::Global(GlobalIndex::from_u32(global_idx)),
-			) {
-				Export::Global(ExportGlobal { definition, global, .. }) => (definition, global),
-				_ => unreachable!("only globals can be returned for a global request"),
-			};
-
-			// skip immutable globals.
-			if !global.mutability {
-				continue;
-			}
-
-			let value = unsafe {
-				// Safety of this function solely depends on the correctness of the reference and
-				// the type information of the global.
-				read_global(def, global.ty)?
-			};
-			preserved_mut_globals.push((def, value));
-		}
-
-		Ok(Self {
-			preserved_mut_globals,
-			handle,
-		})
+		let data = instance_wrapper.instance
+			.exports()
+			.enumerate()
+			.filter_map(|(index, export)| {
+				if export.name().starts_with("exported_internal_global") {
+					export.into_global().map(
+						|g| SavedValue { index, value: into_value(g.get()) }
+					)
+				} else { None }
+			})
+			.collect::<Vec<_>>();
+
+		Ok(Self(data))
 	}
 
 	/// Apply the snapshot to the given instance.
 	///
 	/// This instance must be the same that was used for creation of this snapshot.
 	pub fn apply(&self, instance_wrapper: &InstanceWrapper) -> Result<()> {
-		if instance_wrapper.instance.handle().handle != self.handle {
-			return Err(Error::from("unexpected instance handle".to_string()));
-		}
-
-		for (def, value) in &self.preserved_mut_globals {
-			unsafe {
-				// The following writes are safe if the precondition that this is the same instance
-				// this snapshot was created with:
-				//
-				// 1. These pointers must be still not-NULL and allocated.
-				// 2. The set of global variables is fixed for the lifetime of the same instance.
-				// 3. We obviously assume that the wasmtime references are correct in the first place.
-				// 4. We write the data with the same type it was read in the first place.
-				write_global(*def, *value)?;
+		let mut current = 0;
+		for (index, export) in instance_wrapper.instance.exports().enumerate() {
+			if current >= self.0.len() { break; }
+			let current_saved = &self.0[current];
+			if index < current_saved.index { continue; }
+			else if index > current_saved.index { current += 1; continue; }
+			else {
```

Comment:
TBH I am not sure what we are doing here.  

Can't we just iterate over `self.0` and then set the item under the corresponding index from `exports`? It is an iterator, but `nth` should work and since it is `ExactSizeIterator` it should be efficient enough.

Alternatively, since a `Global` is essentially ref-counted smart-pointer, we could store it in `SavedValue` and avoid accessing `exports` at all.

Reply:
What we do here is essentiallly move two pointers over ordered lists together to see when they have equal values. I am not sure that `nth` won't iterate over all iterator every time even for `ExactSizeIterator` (methods of this trait have nothing about this). Can you provide proof of that? :)

Global is not `Sync`/`Send`, so I'd avoid saving it anywhere if possible.

Reply:
yeah ok don't think this is worth arguing about. The comments improve the situation as well.